File: queues/ibmmq/Python/ibmmqclient.py
"""
Copyright Core DF

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

IBM MQ helpers for Core Auto. **Put** from step scripts; **Get** for
[ingress](../../ingress/Python/README.md) bridges only (not steps).
"""

import json
import os
from typing import Any, Optional

from lib.result import missing_env, transport_error


def _queue_name(explicit: Optional[str]) -> str:
    return explicit or os.environ.get("MQ_QUEUE", "")


def _encode(value: Any) -> bytes:
    if isinstance(value, bytes):
        return value
    if isinstance(value, (dict, list)):
        return json.dumps(value).encode("utf-8")
    return str(value).encode("utf-8")


def _decode(raw: bytes) -> Any:
    if isinstance(raw, str):
        raw = raw.encode("utf-8")
    try:
        return json.loads(raw.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError):
        return raw.decode("utf-8", errors="replace")


def _connect():
    import pymqi

    host = os.environ.get("MQ_HOST", "")
    port = os.environ.get("MQ_PORT", "1414")
    qmgr_name = os.environ.get("MQ_QUEUE_MANAGER", "")
    channel = os.environ.get("MQ_CHANNEL", "SYSTEM.DEF.SVRCONN")
    user = os.environ.get("MQ_USER", "")
    password = os.environ.get("MQ_PASSWORD", "")

    if not host or not qmgr_name:
        raise ValueError("MQ_HOST and MQ_QUEUE_MANAGER required")

    conn_info = f"{host}({port})"
    cd = pymqi.CD()
    cd.ChannelName = channel.encode()
    cd.ConnectionName = conn_info.encode()
    cd.ChannelType = pymqi.CMQC.MQCHT_CLNT
    cd.TransportType = pymqi.CMQC.MQXPT_TCP

    if user:
        return pymqi.connect(qmgr_name, cd, user, password)
    return pymqi.connect(qmgr_name, cd)


def Init() -> dict:
    if not os.environ.get("MQ_HOST") or not os.environ.get("MQ_QUEUE_MANAGER"):
        return missing_env("MQ_HOST and MQ_QUEUE_MANAGER")
    if not os.environ.get("MQ_QUEUE"):
        return missing_env("MQ_QUEUE (or pass queue per call)")
    return {"status_code": 200}


def Put(value: Any, queue: Optional[str] = None) -> dict:
    qname = _queue_name(queue)
    if not qname:
        return missing_env("MQ_QUEUE")
    try:
        import pymqi

        qmgr = _connect()
        q = pymqi.Queue(qmgr, qname)
        try:
            q.put(_encode(value))
        finally:
            q.close()
            qmgr.disconnect()
        return {"status_code": 200}
    except ImportError:
        return {"status_code": 500, "error": "pymqi package required (IBM MQ client libraries must be installed)"}
    except Exception as exc:
        return transport_error(str(exc))


def Get(
    queue: Optional[str] = None,
    timeout_sec: float = 30,
    max_messages: int = 1,
) -> dict:
    """Get messages from a queue (wait). For ingress processes only — not step scripts."""
    qname = _queue_name(queue)
    if not qname:
        return missing_env("MQ_QUEUE")
    try:
        import pymqi

        qmgr = _connect()
        q = pymqi.Queue(qmgr, qname)
        gmo = pymqi.GMO()
        gmo.Options = pymqi.CMQC.MQGMO_WAIT | pymqi.CMQC.MQGMO_NO_SYNCPOINT
        gmo.WaitInterval = int(timeout_sec * 1000)
        messages = []
        try:
            for _ in range(max(1, max_messages)):
                try:
                    msg = q.get(None, pymqi.MD(), gmo)
                except pymqi.MQMIError as err:
                    if err.reason == pymqi.CMQC.MQRC_NO_MSG_AVAILABLE:
                        break
                    raise
                messages.append({"queue": qname, "value": _decode(msg)})
        finally:
            q.close()
            qmgr.disconnect()
        return {"status_code": 200, "messages": messages}
    except ImportError:
        return {"status_code": 500, "error": "pymqi package required (IBM MQ client libraries must be installed)"}
    except Exception as exc:
        return transport_error(str(exc))

File: queues/ibmmq/Rust/src/ibmmqclient.rs
// Copyright Core DF — Apache License 2.0

use crate::result::{missing_env, transport_error};
use base64::{engine::general_purpose::STANDARD as B64, Engine};
use serde_json::{json, Value};
use std::env;

fn queue_name(explicit: Option<&str>) -> String {
    explicit
        .filter(|s| !s.is_empty())
        .map(str::to_string)
        .or_else(|| env::var("MQ_QUEUE").ok())
        .unwrap_or_default()
}

fn base_url() -> String {
    if let Ok(v) = env::var("MQ_REST_BASE_URL") {
        if !v.is_empty() {
            return v.trim_end_matches('/').to_string();
        }
    }
    let host = env::var("MQ_HOST").unwrap_or_default();
    let port = env::var("MQ_REST_PORT").unwrap_or_else(|_| "9443".into());
    format!("https://{host}:{port}/ibmmq/rest/v2")
}

fn auth_header() -> Option<String> {
    let user = env::var("MQ_USER").unwrap_or_default();
    let password = env::var("MQ_PASSWORD").unwrap_or_default();
    if user.is_empty() {
        return None;
    }
    Some(format!("Basic {}", B64.encode(format!("{user}:{password}"))))
}

pub fn init() -> Value {
    if env::var("MQ_HOST").unwrap_or_default().is_empty() || env::var("MQ_QUEUE_MANAGER").unwrap_or_default().is_empty() {
        return missing_env("MQ_HOST and MQ_QUEUE_MANAGER");
    }
    if env::var("MQ_QUEUE").unwrap_or_default().is_empty() {
        return missing_env("MQ_QUEUE (or pass queue per call)");
    }
    json!({"status_code":200})
}

pub fn put(value: Value, queue: Option<&str>) -> Value {
    let qname = queue_name(queue);
    if qname.is_empty() {
        return missing_env("MQ_QUEUE");
    }
    let qmgr = env::var("MQ_QUEUE_MANAGER").unwrap_or_default();
    let body_str = if value.is_string() {
        value.as_str().unwrap_or("").to_string()
    } else {
        value.to_string()
    };
    let url = format!("{}/messaging/qmgr/{qmgr}/queue/{qname}/message", base_url());
    let payload = json!({"type":"string","content": body_str});
    let mut req = ureq::post(&url).set("Content-Type", "application/json");
    if let Some(auth) = auth_header() {
        req = req.set("Authorization", &auth);
    }
    match req.send_json(payload) {
        Ok(resp) if resp.status() >= 400 => json!({"status_code": resp.status(), "error": resp.into_string().unwrap_or_default()}),
        Ok(_) => json!({"status_code":200}),
        Err(ureq::Error::Status(code, resp)) => json!({"status_code":code,"error":resp.into_string().unwrap_or_default()}),
        Err(e) => transport_error(&e.to_string()),
    }
}

pub fn get(queue: Option<&str>, _timeout_sec: f64, max_messages: usize) -> Value {
    let qname = queue_name(queue);
    if qname.is_empty() {
        return missing_env("MQ_QUEUE");
    }
    let qmgr = env::var("MQ_QUEUE_MANAGER").unwrap_or_default();
    let mut messages = Vec::new();
    for _ in 0..max_messages.max(1) {
        let url = format!("{}/messaging/qmgr/{qmgr}/queue/{qname}/message", base_url());
        let mut req = ureq::delete(&url);
        if let Some(auth) = auth_header() {
            req = req.set("Authorization", &auth);
        }
        let resp = match req.call() {
            Ok(r) => r,
            Err(ureq::Error::Status(204, _)) => break,
            Err(e) => return transport_error(&e.to_string()),
        };
        if resp.status() >= 400 {
            return json!({"status_code": resp.status(), "error": resp.into_string().unwrap_or_default()});
        }
        let text = resp.into_string().unwrap_or_default();
        let parsed: Value = serde_json::from_str(&text).unwrap_or(Value::String(text));
        let content = parsed.get("content").cloned().unwrap_or(parsed);
        let value = if content.is_string() {
            serde_json::from_str(content.as_str().unwrap_or("")).unwrap_or(content)
        } else {
            content
        };
        messages.push(json!({"queue": qname, "value": value}));
    }
    json!({"status_code":200,"messages":messages})
}
